### DISCUSSION TITLE
[draft/discussing] 拆分日志地址、插件地址、配置文件地址

### DISCUSSION THREAD

**Discussion by pizeroLOL** [Class-Widgets/Class-Widgets#306] [Feature]
## 背景与动机

1. 方便 #180 实现
2. 允许 Linux 开发者将该软件添加到软件仓库

## 想要实现或优化的功能

- [ ] macOS、Windows 相关规范待补充
- [ ] 等待插件相关建议补充

### 环境变量

> **Tip**
> ClassWidgets 全平台通用

- `CLASSWIDGETS_NOT_PORTABLE`：用于判断是否为便携版
- `CLASSWIDGETS_CUSTOM_LOG_HOME`：用于用户自定义的日志位置
- `CLASSWIDGETS_CUSTOM_CONFIG_HOME`：用于用户自定义的配置存放位置
- `CLASSWIDGETS_CUSTOM_PLUGINS_HOME`：用于存放用户自定义的插件存放位置

> **Tip**
> unix like 或者叫支持了 [XDG 相关规范](https://specifications.freedesktop.org/basedir-spec/latest/) 的平台使用

- `XDG_CONFIG_HOME`：用户可见的配置存放位置
- `XDG_STATE_HOME`: 用户应用状态存放位置
- `XDG_CACHE_HOME`：用户 log 存放位置

> **Tip**
> unix like 通常支持，无文档，但是是惯例。

- `HOME`：用户家目录

> **Tip**
> 
> NT 系统，或者叫 Windows 特例通常支持，暂时找不到文档，等个好心人。

- `APPDATA`：用户应用数据存放地址
- `TEMP`：用户临时文件存放位置
- `USERPROFILE`：用户文件夹

### 特殊地址

macOS 这边可能的参考文档：https://developer.apple.com/documentation/foundation/filemanager/searchpathdirectory/applicationsupportdirectory

- `~/Library/Application Support`：苹果应用支持目录，用例：https://ghostty.org/docs/config
- `~/Library/Cache`：苹果临时文件地址，等人补充用例

Windows 这边还有个 `LocalLow`，就在 `%APPDATA%\..\LocalLow` 目前我这台电脑上看到的用例就是只有 Intel shader 缓存和 shapez 2 的游戏存档。

### 目前想法

1. 我不是很清楚目前插件存放位置的实现，如果实现是把列表写到 config 文件夹里，然后通过网络下载，那么这东西和 cache 一桌（就扔日志的那个文件夹，反正到时候会自己下回来）；如果只是把文件扔到哪个指定目录，那么这东西和 state 一桌（`XDG_STATE_HOME`，Windows 这边可以用 `LocalLow`）。
2. 如果不想搞这么杂七杂八的，只要判断一下是不是便携版，然后扔环境变量里，是便携版就扔 `./config`、`./log`、`./plugin`，不是就扔 `$HOME/.classwidgets/<上面的相对路径>` 或者 `%USERPROFILE%\.classwidgets` 顺手搞个隐藏。

## 考虑实现

所有标明 `没有则创建` 的为最终匹配

### 配置文件

0. `./config`：如果 `CLASSWIDGETS_NOT_PORTABLE` 不存在或为空
1. `<CLASSWIDGETS_CUSTOM_CONFIG_HOME>`：如果环境变量已设置且不为空，没有则创建
2. `%APPDATA%\Class Widgets\config`：如果平台为 Windows，没有则创建
3. `$XDG_CONFIG_HOME/Class Widgets/config`：如果平台为不为 Windows 和 macOS，且环境变量设置且不为空，没有则创建
4. `$HOME/.config/Class Widgets/config`：XDG config home 默认实现，如果平台不为 Windows 和 macOS，环境变量没置或为空直接报错，没有文件夹则创建
5. `$HOME/Library/Application Support/Class Widgets/config`：macOS 平台环境变量没设置或为空直接报错，没有则创建

### log

0. `./log`：如果 `CLASSWIDGETS_NOT_PORTABLE` 不存在或为空
1. `<CLASSWIDGETS_CUSTOM_LOG_HOME>`：如果环境变量已设置且不为空，没有则创建
2. `%TMP%\Class Widgets\log`：如果平台为 Windows，没有则创建
3. `$XDG_CACHE_HOME/Class Widgets/log`：如果平台为不为 Windows 和 macOS，且环境变量设置且不为空，没有则创建
4. `$HOME/.cache/Class Widgets`：XDG cache home 默认实现，如果平台不为 Windows 和 macOS，环境变量没置或为空直接报错，没有文件夹则创建
5. `$HOME/Library/Cache/Class Widgets/log`：macOS 平台环境变量没设置或为空直接报错，没有则创建

### 插件

感谢 @xuanxuan1231 

0. `./plugins`：如果 `CLASSWIDGETS_NOT_PORTABLE` 不存在或为空
1. `<CLASSWIDGETS_CUSTOM_PLUGIN_HOME>`：如果环境变量已设置且不为空，没有则创建
2. `%APPDATA%\Class Widgets\plugins`：如果平台为 Windows，没有则创建
3. `$XDG_DATA_HOME/Class Widgets/plugins`：如果平台为不为 Windows 和 macOS，且环境变量设置且不为空，没有则创建
4. `$HOME/.local/share/Class Widgets/plugins`：如果平台不为 Windows 和 macOS，环境变量没置或为空直接报错，没有文件夹则创建
5. `$HOME/Library/Application Support/Class Widgets/plugins`：macOS 平台环境变量没设置或为空直接报错，没有则创建

**Comment by xuanxuan1231** [comment]
cw的所有依赖文件好像都是以相对路径引用的，所以调工作目录应该就可以达到这样的效果（？）

**Comment by xuanxuan1231** [comment]
现在已经用绝对路径了，所以说只要让 base_directory 用上面那两个就可以了

**Comment by xuanxuan1231** [comment]
@pizeroLOL @RinLit-233-shiroko 

## 调整

> ./config：如果 CLASSWIDGETS_IS_PORTABLE 为 true、TRUE、1

考虑改为是否存在，应该还要更个名 因为安装程序可以创建环境变量。下同。

> $HOME/Library/Application Support/io.github.class-widgets.class-widgets/config：macOS 平台环境变量没设置或为空直接报错，没有则创建

也可以改为 `~/Library/Application Support/Class Widgets/config`。下同。

## 补充
关于插件：

0. `./plugins`：如果 CLASSWIDGETS_IS_PORTABLE 不存在
1. `<CLASSWIDGETS_CUSTOM_PLUGIN_HOME>`：如果环境变量已设置且不为空，没有则创建
2. `%APPDATA%\ClassWidgets\plugins`：如果平台为 Windows，没有则创建
4.  `$XDG_DATA_HOME/ClassWidgets/plugins`：如果平台为不为 Windows 和 macOS，且环境变量设置且不为空，没有则创建
5. 待定：如果平台不为 Windows 和 macOS，环境变量没置或为空直接报错，没有文件夹则创建
6. `$HOME/Library/Application Support/Class Widgets/plugins`：macOS 平台环境变量没设置或为空直接报错，没有则创建
